### DISCUSSION TITLE
YOLOv5 dataset 사용법

### DISCUSSION THREAD

**Discussion by jaehwan-AI** [boostcampaitech2/object-detection-level2-cv-06#106] [ideas]
**_YOLOv5 데이터셋은 원하시면 따로 공유드리겠습니다._**

<YOLOv5 사용법>
[yolov5 github](https://github.com/ultralytics/yolov5)에 가보시면 사용법이 나와있지만 우리 데이터셋에 맞는 사용법을 간단히 공유해봅니다.
![image](https://user-images.githubusercontent.com/66052001/136378353-827a41f2-fd73-4fad-8ae5-2f675c8c731c.png)
yolov5를 git clone해서 받으시면 다음과 같이 clone이 됩니다.

**빨간 네모는 제가 실행하면서 생긴 것들이라 clone 받아도 없습니다.**

이제부터 yolov5 간편 사용 설명 시작하겠습니다.

첫 번째로 빨간 네모들 부터 설명드릴게요
1. dataset : 실제로 사용하는 dataset을 모아놓았는데 다른 경로가 있다면 그 곳으로 옮겨도 상관없습니다.
   대신 구조는 아래와 같이 해주셔야 합니다.
   ![image](https://user-images.githubusercontent.com/66052001/136379122-720398e2-5e2a-466a-b636-bec95e5f9c07.png)
2. runs는 학습이나 inference하면 자동으로 결과가 담기는 폴더입니다. 현재 mmdetection의 work_dirs라고 생각하시면 됩니다.
3. wandb는 yolov5랑 wandb 연동했을 때 생기는 폴더입니다.
4. inference.py는 yolov5를 돌리고 현재 대회 submission하기 위해 만든 파일입니다. 이건 필요하시면 따로 설명드릴게요.
5. mk_trainset.py랑 mk_testset.py는 yolo용 학습 데이터 만들 때 사용했던 건데 이제는 라이브러리가 있으니 필요가 없습니다...
6. yolov5s.pt와 yolov5x.pt는 pretrained weight인데 나중에 학습이나 inference할 때 자동으로 install 됩니다.

이제 사용법 설명드릴게요
1. 아래 사진과 같이 yaml 파일을 만들어서 data 라는 폴더에 넣어주세요.
    ![image](https://user-images.githubusercontent.com/66052001/136380055-0817cef3-620f-46f8-9799-17584a54b839.png)
    path에 아까 말씀드린 dataset이 들어가 있는 것을 확인할 수 있습니다.
2. 그리고 models 폴더에 가보시면 모델 사이즈별로 s, m, l, x 4종류가 있는데 원하시는 것을 복사하셔서 아래와 같이 class 수만
    변경해주시면 학습 준비가 끝납니다. 각 모델 첫 줄에 nc라고 있습니다.
    ![image](https://user-images.githubusercontent.com/66052001/136380390-4f685506-1fb3-4d64-860f-5e07c3b22db3.png)
3. yolov5 폴더 경로로 이동한 다음 다음과 같이 입력하면 바로 학습이 진행됩니다.
    `python train.py --img-size 1024 --batch {batch_size} --epochs {epochs} --data ./data/{1번에서 만든 yaml} --cfg {2번에서 만든 모델 yaml 파일} --weights {선택한 yolov5 모델, ex) yolov5s.pt} --entity {wandb entity} --project {wandb project} --name {wandb project/name}`

저는 간단하게 학습하고 inference하는 방법만 알고 다른 세세한 설정 부분은 아직 모릅니다.
혹시 막히는 부분을 말씀해주시면 같이 고민해드리겠습니다!

**Comment by jiwoo0212** [comment]
세심한 튜토리얼 감사합니다!

**Comment by ppskj178** [comment]
`dataset : 실제로 사용하는 dataset을 모아놓았는데 다른 경로가 있다면 그 곳으로 옮겨도 상관없습니다.
대신 구조는 아래와 같이 해주셔야 합니다.`
이 작업할떄 실제 데이터를 복사하면 매우 시간이 많이 걸릴 것 같은데
리눅스에서 제공하는 심볼릭 링크를 사용하시면 편하실거에요

![image](https://user-images.githubusercontent.com/20790778/136494118-ca201091-ae83-4d63-8920-f230e9a0ea08.png)

현재 제가 efficientdet 테스트를 위해 구성한 디렉토리인데
저 데이터를 해당 경로로 복사한 것이 아니라 원본 dataset 폴더의 바로가기이고
내부에 valid 폴더 역시 저희는 따로 분할은 안했기때문에 train폴더의 바로가기를 생성했습니다.

명령어는 `ln` 이고 폴더에 사용하실거면 -s 옵션을 주시면됩니다.
